control/controlclient/noise.go

```diff
@@ -0,0 +1,162 @@
+	c.pool.connClosed(c.id)
+	return nil
+}
+
+// noiseClient provides a http.Client to connect to tailcontrol over
+// the ts2021 protocol.
+type noiseClient struct {
+	*http.Client // HTTP client used to talk to tailcontrol
+	privKey      key.MachinePrivate
+	serverPubKey key.MachinePublic
+	connCount    int
+
+	serverURL  string // the full server url with scheme (no trailing slash)
+	serverHost string // the host:port part of serverURL
+
+	mu       sync.Mutex
+	connPool map[int]*noiseConn // All created connections. TODO(maisem): clean up dead connections.
+}
+
+// newNoiseClient returns a new noiseClient for the provided server and machine key.
+// serverURL is of the form https://<host>:<port> (no trailing slash).
+func newNoiseClient(ctx context.Context, priKey key.MachinePrivate, serverURL string) (*noiseClient, error) {
+	u, err := url.Parse(serverURL)
+	if err != nil {
+		return nil, err
+	}
+	np := &noiseClient{
+		privKey:    priKey,
+		serverURL:  serverURL,
+		serverHost: u.Host,
+	}
+
+	// Create a new http.Client which dials out using nc.Dial.
+	np.Client = &http.Client{
+		Transport: &http2.Transport{
+			DialTLS:   np.Dial,
+			AllowHTTP: true,
+		},
+	}
+
+	return np, nil
+}
+
+// connClosed is called from noiseConn.Close to clean up the connection
+// from the pool.
+func (nc *noiseClient) connClosed(id int) {
+	nc.mu.Lock()
+	defer nc.mu.Unlock()
+	delete(nc.connPool, id)
+}
+
+// Close closes all the underlying noise connections.
+// It is a no-op and returns nil if the connection is already closed.
+func (nc *noiseClient) Close() error {
+	nc.mu.Lock()
+	conns := nc.connPool
+	nc.connPool = nil
+	nc.mu.Unlock()
+
+	var errors []error
+	for _, c := range conns {
+		if err := c.Close(); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return multierr.New(errors...)
+}
+
+// Dial opens a new connection to tailcontrol, fetching the server noise key
+// if not cached. The network and address params are ignored as it only dials
+// the server it was created for.
+func (nc *noiseClient) Dial(_, _ string, cfg *tls.Config) (net.Conn, error) {
+	nc.mu.Lock()
+	priKey := nc.privKey
+	serverKey := nc.serverPubKey
+	serverURL := nc.serverURL
+	serverHost := nc.serverHost
+	connID := nc.connCount
+	if nc.connPool == nil {
+		nc.connPool = make(map[int]*noiseConn)
+	}
+	nc.connCount++
+	nc.mu.Unlock()
+
+	// This context is only used for the dial.
+	// Pick an arbitrary timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	if serverKey.IsZero() {
+		sk, err := loadServerNoiseKey(ctx, http.DefaultClient, serverURL)
+		if err != nil {
+			return nil, err
+		}
+		nc.mu.Lock()
+		if nc.serverPubKey.IsZero() {
+			serverKey = sk
+			nc.serverPubKey = serverKey
+		} else {
+			// We lost the race, use the key that won.
+			serverKey = nc.serverPubKey
+		}
+		nc.mu.Unlock()
+	}
+
+	conn, err := controlhttp.Dial(ctx, serverHost, priKey, serverKey)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
this is kind of annoying. We need to give `controlhttp.Dial` an addr which goes over the http endpoint.
https://github.com/tailscale/tailscale/blob/0e62a7d1a247eb312f340b08e92942f73149f11f/control/controlhttp/client.go#L58-L82